cmd/proxy/main.go

```diff
@@ -0,0 +1,178 @@
+			Role:                         conf.Vault.Role,
+			Mount:                        conf.Vault.Mount,
+			TimeToLive:                   conf.Vault.TimeToLive,
+			URISubjectAlternativeNames:   conf.Vault.URISubjectAlternativeNames,
+			OtherSubjectAlternativeNames: conf.Vault.OtherSubjectAlternativeNames,
+			TLSConfig:                    &tls.Config{},
+		}
+		if conf.Vault.CACertPath != "" {
+			v.TLSConfig.RootCAs = x509.NewCertPool()
+			bts, err := ioutil.ReadFile(conf.Vault.CACertPath)
+			if err != nil {
+				log.WithError(err).Fatal("Failed to read CA cert file.")
+			}
+			v.TLSConfig.RootCAs.AppendCertsFromPEM(bts)
+		}
+		issuer = v
+	case envtypes.CFSSLIssuer:
+		if conf.CFSSL.URL.String() == "" {
+			log.Fatal("CFSSL URL is required")
+		}
+		c := &cfssl.Issuer{
+			URL:       &conf.CFSSL.URL,
+			Profile:   conf.CFSSL.Profile,
+			TLSConfig: &tls.Config{},
+		}
+		if conf.CFSSL.CACertPath != "" {
+			c.TLSConfig.RootCAs = x509.NewCertPool()
+			bts, err := ioutil.ReadFile(conf.Vault.CACertPath)
+			if err != nil {
+				log.WithError(err).Fatal("Failed to read CA cert file.")
+			}
+			c.TLSConfig.RootCAs.AppendCertsFromPEM(bts)
+		}
+		if conf.CFSSL.AuthKey != "" {
+			st, err := auth.New(conf.CFSSL.AuthKey, nil)
+			if err != nil {
+				log.WithError(err).Fatal("Failed to create CFSSL auth.")
+			}
+			c.Auth = st
+		}
+		issuer = c
+	case envtypes.AWSIssuer:
+		if conf.AWS.Region == "" {
+			log.Fatal("AWS region is required")
+		}
+		if conf.AWS.AccessKeyID == "" {
+			log.Fatal("AWS access key ID is required")
+		}
+		if conf.AWS.AccessKeySecret == "" {
+			log.Fatal("AWS access key secret is required")
+		}
+		if conf.AWS.CertificateAuthorityARN == "" {
+			log.Fatal("AWS CA ARN is required")
+		}
+		ac := awssdk.Config{}
+		ac.Region = conf.AWS.Region
+		ac.Credentials = awssdk.NewStaticCredentialsProvider(conf.AWS.AccessKeyID, conf.AWS.AccessKeySecret, "")
+		issuer = &aws.Issuer{
+			Client:                  acmpca.New(ac),
+			CertificateAuthorityARN: conf.AWS.CertificateAuthorityARN,
+			TimeToLive:              conf.AWS.TimeToLive,
+		}
+	}
+
+	cache := certify.NewMemCache()
+	if conf.CacheDir != "" {
+		cache = certify.DirCache(conf.CacheDir)
+	}
+
+	c := certify.Certify{
+		Issuer: issuer,
+		Cache:  cache,
+		Logger: logrusadapter.New(log),
+
+		CommonName:   conf.CommonName,
+		RenewBefore:  conf.RenewBefore,
+		IssueTimeout: conf.IssueTimeout,
+		CertConfig: &certify.CertConfig{
+			SubjectAlternativeNames:   conf.SubjectAlternativeNames,
+			IPSubjectAlternativeNames: conf.IPSubjectAlternativeNames,
+		},
+	}
+
+	u := &url.URL{
+		Scheme: "http",
+		Host:   conf.ProxyAddr,
+	}
+
+	s := &http.Server{
+		Addr:    conf.Addr,
+		Handler: httputil.NewSingleHostReverseProxy(u),
+		TLSConfig: &tls.Config{
+			GetCertificate:       c.GetCertificate,
+			GetClientCertificate: c.GetClientCertificate,
+		},
+	}
+
+	log.Infof("Proxying traffic from %q to %q", conf.Addr, conf.ProxyAddr)
+
+	err = s.ListenAndServeTLS("", "")
```

Comment:
Would be nice to shutdown the server when receiving SIGTERM or SIGINT.

Reply:
Good idea!